egonlang-verifier/src/rules/core/assert_type.rs

```diff
@@ -2,20 +2,26 @@ use crate::prelude::*;
 use egonlang_core::ast::Stmt;
 use egonlang_errors::{EgonErrorS, EgonTypeError};
 
-stmt_rule!(AssertType, |stmt, span, _resolve_ident, resolve_expr| {
+stmt_rule!(AssertType, |stmt, _span, _resolve_ident, resolve_expr| {
     let mut errs: Vec<EgonErrorS> = vec![];
 
-    if let Stmt::AssertType(x) = stmt {
-        if let Ok(value_type) = resolve_expr(&x.value.0, &x.value.1) {
-            if let Ok(expected_type_type) = resolve_expr(&x.expected_type.0, &x.expected_type.1) {
-                if value_type.of_type != expected_type_type.of_type {
+    if let Stmt::AssertType(stmt_assert_type) = stmt {
+        if let Some(value_type) = resolve_expr.get(&(
+            stmt_assert_type.value.0.clone(),
+            stmt_assert_type.value.1.clone(),
+        )) {
```

Comment:
`resolve_expr` now takes both Expr and Span to look up expr type from the expr type cache